### PR TITLE
DAY_TO_STR() localization

### PR DESCRIPTION
**Pull request by alejandro-colomar**
Cc: @eggert 
Cc: @timparenti
Cc: @kenion

---

Revisions:

<details>
<summary>v2</summary>

-  Print localized dates and times by default.
-  Do not use custom formats.
-  Print timezone information together with date strings, as an RFC 9557 suffix.
-  Fix STRFTIME() macro

```
$ git range-diff shadow/master gh/time time 
1:  2686a033 < -:  -------- src/chage.c: Always use ISO 8601 format (YYYY-MM-DD) for printing dates
-:  -------- > 1:  68a97728 lib/, src/: Use %F instead of %Y-%m-%d with strftime(3)
-:  -------- > 2:  56067591 src/: Use %c for printing localized date&times with strftime(3)
-:  -------- > 3:  928172c4 src/chage.c: print_day_as_date(): Use %x for printing localized dates with strftime(3)
2:  de6f93bf = 4:  e78f9782 src/chage.c: print_day_as_date(): Simplify error handling
-:  -------- > 5:  fc46d1f3 lib/time/, src/: day_to_str(): Add support for formatting localized dates
3:  bd7701fa ! 6:  b2c33c58 src/chage.c: print_day_as_date(): Call DAY_TO_STR() to de-duplicate code
    @@ src/chage.c: static int new_fields (void)
     -          return;
     -  }
     -
    --  STRFTIME(buf, "%Y-%m-%d", &tm);
    +-  STRFTIME(buf, iflg ? "%F" : "%x", &tm);
     -  (void) puts (buf);
    -+  DAY_TO_STR(buf, day);
    ++  DAY_TO_STR(buf, day, iflg);
     +  puts(buf);
      }
      
-:  -------- > 7:  9c0ee33c lib/time/day_to_str.h: day_to_str(): Print numeric timezone together with local dates
-:  -------- > 8:  41a12dfe lib/string/strftime.h: STRFTIME(): Tighten macro definition
```
</details>

<details>
<summary>v3</summary>

-  Use the time zone name/abbreviation rather than the numeric offset.  It's strongly preferred by RFC 9557.  [@eggert]

```
$ git range-diff shadow/master gh/time time 
1:  68a97728 = 1:  68a97728 lib/, src/: Use %F instead of %Y-%m-%d with strftime(3)
2:  56067591 = 2:  56067591 src/: Use %c for printing localized date&times with strftime(3)
3:  928172c4 = 3:  928172c4 src/chage.c: print_day_as_date(): Use %x for printing localized dates with strftime(3)
4:  e78f9782 = 4:  e78f9782 src/chage.c: print_day_as_date(): Simplify error handling
5:  fc46d1f3 = 5:  fc46d1f3 lib/time/, src/: day_to_str(): Add support for formatting localized dates
6:  b2c33c58 = 6:  b2c33c58 src/chage.c: print_day_as_date(): Call DAY_TO_STR() to de-duplicate code
7:  9c0ee33c ! 7:  db4f966b lib/time/day_to_str.h: day_to_str(): Print numeric timezone together with local dates
    @@ Metadata
     Author: Alejandro Colomar <alx@kernel.org>
     
      ## Commit message ##
    -    lib/time/day_to_str.h: day_to_str(): Print numeric timezone together with local dates
    +    lib/time/day_to_str.h: day_to_str(): Print timezone together with local dates
     
         Print it as a RFC 9557 suffix, which doesn't add any extra whitespace.
         This reduces the damage to expectations of scripts that might parse
    @@ Commit message
         Link: <https://github.com/shadow-maint/shadow/issues/1057>
         Link: <https://lists.sr.ht/~hallyn/shadow/%3Cmzbl4gkns5nyt7otg4t7cxlds2q64wutpjnngxpnix55ocbds2@whvswk343dwl%3E>
         Link: <https://www.rfc-editor.org/rfc/rfc9557.pdf>
    +    Link: <https://lists.gnu.org/archive/html/bug-gnulib/2024-08/msg00003.html>
         Suggested-by: Alejandro Colomar <alx@kernel.org>
         Suggested-by: Paul Eggert <eggert@cs.ucla.edu>
         Cc: "Serge E. Hallyn" <serge@hallyn.com>
         Cc: Tim Parenti <tim@timtimeonline.com>
         Cc: Gus Kenion <https://github.com/kenion>
         Cc: Iker Pedrosa <ipedrosa@redhat.com>
    +    Cc: Bruno Haible <bruno@clisp.org>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## lib/time/day_to_str.h ##
    @@ lib/time/day_to_str.h: day_to_str(size_t size, char buf[size], long day, bool is
        }
      
     -  if (strftime(buf, size, iso ? "%F" : "%x", &tm) == 0)
    -+  if (strftime(buf, size, iso ? "%F[%z]" : "%x[%z]", &tm) == 0)
    ++  if (strftime(buf, size, iso ? "%F[%Z]" : "%x[%Z]", &tm) == 0)
                strtcpy(buf, "future", size);
      }
      
8:  41a12dfe = 8:  bf3481df lib/string/strftime.h: STRFTIME(): Tighten macro definition
```
</details>

<details>
<summary>v4</summary>

-  Don't use a RFC 9557 suffix; that is, don't print timezone info.  date(1) doesn't yet support them, so let's wait until that happens.

```
$ git range-diff shadow/master gh/time time 
1:  68a97728 = 1:  68a97728 lib/, src/: Use %F instead of %Y-%m-%d with strftime(3)
2:  56067591 = 2:  56067591 src/: Use %c for printing localized date&times with strftime(3)
3:  928172c4 = 3:  928172c4 src/chage.c: print_day_as_date(): Use %x for printing localized dates with strftime(3)
4:  e78f9782 = 4:  e78f9782 src/chage.c: print_day_as_date(): Simplify error handling
5:  fc46d1f3 = 5:  fc46d1f3 lib/time/, src/: day_to_str(): Add support for formatting localized dates
6:  b2c33c58 = 6:  b2c33c58 src/chage.c: print_day_as_date(): Call DAY_TO_STR() to de-duplicate code
7:  db4f966b < -:  -------- lib/time/day_to_str.h: day_to_str(): Print timezone together with local dates
8:  bf3481df = 7:  62b553e3 lib/string/strftime.h: STRFTIME(): Tighten macro definition
```
</details>

<details>
<summary>v5</summary>

-  Rebase

```
$ git range-diff master..gh/time shadow/master..time
1:  68a97728 < -:  -------- lib/, src/: Use %F instead of %Y-%m-%d with strftime(3)
2:  56067591 = 1:  b5d8ff0e src/: Use %c for printing localized date&times with strftime(3)
3:  928172c4 < -:  -------- src/chage.c: print_day_as_date(): Use %x for printing localized dates with strftime(3)
4:  e78f9782 < -:  -------- src/chage.c: print_day_as_date(): Simplify error handling
5:  fc46d1f3 = 2:  53554dc6 lib/time/, src/: day_to_str(): Add support for formatting localized dates
6:  b2c33c58 < -:  -------- src/chage.c: print_day_as_date(): Call DAY_TO_STR() to de-duplicate code
7:  62b553e3 < -:  -------- lib/string/strftime.h: STRFTIME(): Tighten macro definition
```
</details>

### CODE REVIEW

**Comment by alejandro-colomar** [comment]
Queued after <https://github.com/shadow-maint/shadow/pull/1058>.

**Review by ikerexxe — COMMENTED** [review]
It looks good, I'll take another look when you resolve the conflicts.

**Comment by alejandro-colomar** [comment]
v1b:

-  Rebase

```
$ git range-diff gh/localtime..gh/time shadow/master..time 
1:  6d62fef0 = 1:  2686a033 src/chage.c: Always use ISO 8601 format (YYYY-MM-DD) for printing dates
2:  b48d0342 = 2:  de6f93bf src/chage.c: print_day_as_date(): Simplify error handling
3:  03e91f78 = 3:  bd7701fa src/chage.c: print_day_as_date(): Call DAY_TO_STR() to de-duplicate code
```

File: src/chage.c [hunk 1]
```diff
@@ -247,7 +246,7 @@ print_day_as_date(long day)
 		return;
 	}
 
-	STRFTIME(buf, iflg ? "%Y-%m-%d" : "%b %d, %Y", &tm);
+	STRFTIME(buf, "%Y-%m-%d", &tm);
```
Comment (alejandro-colomar) [inline src/chage.c]: Alternatively, I thought of using `"%c"` instead of `"%b %d, %Y"` in non-ISO mode, and adding a boolean flag to DAY_TO_STR to specify if we want ISO mode or locale mode.

Here's what `"%c"` means (from the :

date(1):
```
%c     locale’s date and time (e.g., Thu Mar  3 23:05:25 2005)
```

strftime(3):
```
     %c     The  preferred date and time representation for the current locale.
            (The specific format used in the current locale can be obtained  by
            calling  nl_langinfo(3) with D_T_FMT as an argument for the %c con‐
            version specification, and with ERA_D_T_FMT for the %Ec  conversion
            specification.)   (In  the POSIX locale this is equivalent to %a %b
            %e %H:%M:%S %Y.)
```

But `"%b %d, %Y"` is something that should be removed, since it's not a valid way of printing dates $here, for many values of $here in the world.

Any opinions?

Comment (alejandro-colomar) [reply src/chage.c]: Ahh, I hadn't found `"%x"`.  It's more appropriate here, as it only includes date information.

date(1):
```
     %x     locale’s date representation (e.g., 12/31/99)
```
strftime(3):
```
     %x     The  preferred  date  representation for the current locale without
            the time.  (The specific format used in the current locale  can  be
            obtained  by  calling  nl_langinfo(3) with D_FMT as an argument for
            the %x conversion specification, and with  ERA_D_FMT  for  the  %Ex
            conversion specification.)  (In the POSIX locale this is equivalent
            to %m/%d/%y.)
```

Comment (alejandro-colomar) [reply src/chage.c]: I'll do that.  We're always in time to switch to always-ISO.  And it's more friendly to use locale for human-readable stuff.

File: src/chage.c [hunk 2]
```diff
@@ -243,7 +243,7 @@ print_day_as_date(long day)
 	}
 
 	if (localtime_r(&date, &tm) == NULL) {
-		(void) printf ("time_t: %lu\n", (unsigned long)date);
+		puts(_("future"));
```
Comment (ikerexxe) [inline src/chage.c]: Why did you change `printf()`?

**Review by ikerexxe — COMMENTED** [review]
Well, this changed a bit since my last review.

I am not completely convinced about the changes, but if we make them we should add in the release notes that from now on all the dates printed follow the ISO standard.

**Comment by alejandro-colomar** [comment]
> Well, this changed a bit since my last review.
> 
> I am not completely convinced about the changes, but if we make them we should add in the release notes that from now on all the dates printed follow the ISO standard.

Nah, I reverted that change.  :)

File: src/chage.c (hunk 2 above)
Comment (alejandro-colomar) [reply src/chage.c]:
```
commit e78f9782183e88e3973d4634d22c538a09f44b96
Author: Alejandro Colomar <alx@kernel.org>
Date:   Wed Jul 31 14:53:12 2024 +0200

    src/chage.c: print_day_as_date(): Simplify error handling
    
    If localtime_r(3) fails, just print future, as we do in day_to_str().
    It should only fail for unrealistic dates, if at all.
    
    This is a step towards reusing the DAY_TO_STR() function.
    
    Signed-off-by: Alejandro Colomar <alx@kernel.org>

diff --git a/src/chage.c b/src/chage.c
index 0f963090..8b2d7b35 100644
--- a/src/chage.c
+++ b/src/chage.c
@@ -243,7 +243,7 @@ print_day_as_date(long day)
        }
 
        if (localtime_r(&date, &tm) == NULL) {
-               (void) printf ("time_t: %lu\n", (unsigned long)date);
+               puts(_("future"));
                return;
        }
 
```

This makes it behave like str_to_day(), so that we can just call it there.

Comment (alejandro-colomar) [reply src/chage.c]:
```diff
$ diff -u <(git show shadow/master:src/chage.c | grepc print_day_as_date) <(git show shadow/master:lib/time/day_to_str.h | grepc day_to_str)
--- /dev/fd/63	2024-08-02 11:49:56.228997826 +0200
+++ /dev/fd/62	2024-08-02 11:49:56.228997826 +0200
@@ -1,25 +1,25 @@
-(standard input):static void print_day_as_date (long day);
-(standard input):static void
-print_day_as_date(long day)
+(standard input):inline void day_to_str(size_t size, char buf[size], long day);
+(standard input):inline void
+day_to_str(size_t size, char buf[size], long day)
 {
-	char       buf[80];
 	time_t     date;
 	struct tm  tm;
 
 	if (day < 0) {
-		puts(_("never"));
+		strtcpy(buf, "never", size);
 		return;
 	}
+
 	if (__builtin_mul_overflow(day, DAY, &date)) {
-		puts(_("future"));
+		strtcpy(buf, "future", size);
 		return;
 	}
 
 	if (localtime_r(&date, &tm) == NULL) {
-		(void) printf ("time_t: %lu\n", (unsigned long)date);
+		strtcpy(buf, "future", size);
 		return;
 	}
 
-	STRFTIME(buf, iflg ? "%Y-%m-%d" : "%b %d, %Y", &tm);
-	(void) puts (buf);
+	if (strftime(buf, size, "%Y-%m-%d", &tm) == 0)
+		strtcpy(buf, "future", size);
 }
```

Comment (alejandro-colomar) [reply src/chage.c]: So, after that commit, the difference is reduced to just:

```diff
$ diff -u \
    <(git show e78f9782183e:src/chage.c \
      | grepc print_day_as_date) \
    <(git show e78f9782183e:lib/time/day_to_str.h \
      | grepc day_to_str);
--- /dev/fd/63	2024-08-02 14:14:36.069372059 +0200
+++ /dev/fd/62	2024-08-02 14:14:36.069372059 +0200
@@ -1,25 +1,25 @@
-(standard input):static void print_day_as_date (long day);
-(standard input):static void
-print_day_as_date(long day)
+(standard input):inline void day_to_str(size_t size, char buf[size], long day);
+(standard input):inline void
+day_to_str(size_t size, char buf[size], long day)
 {
-	char       buf[80];
 	time_t     date;
 	struct tm  tm;
 
 	if (day < 0) {
-		puts(_("never"));
+		strtcpy(buf, "never", size);
 		return;
 	}
+
 	if (__builtin_mul_overflow(day, DAY, &date)) {
-		puts(_("future"));
+		strtcpy(buf, "future", size);
 		return;
 	}
 
 	if (localtime_r(&date, &tm) == NULL) {
-		puts(_("future"));
+		strtcpy(buf, "future", size);
 		return;
 	}
 
-	STRFTIME(buf, iflg ? "%F" : "%x", &tm);
-	(void) puts (buf);
+	if (strftime(buf, size, "%F", &tm) == 0)
+		strtcpy(buf, "future", size);
 }
```

So the only difference after that is the support for localized dates.

**Comment by hallyn** [comment]
So this unfortunately begs the usual question - how many dark server rooms are running 20 year old scripts depending on the current date format that are going to break from a perfectly reasonable change?

I *think* none.  I think this is fine.

**Comment by alejandro-colomar** [comment]
> So this unfortunately begs the usual question - how many dark server rooms are running 20 year old scripts depending on the current date format that are going to break from a perfectly reasonable change?

The two breaking changes are:

<https://github.com/shadow-maint/shadow/pull/1059/commits/56067591034f51c083a347d374ccd0ed6aa3c715>
<https://github.com/shadow-maint/shadow/pull/1059/commits/928172c43495c08b8eb7502cd7890f78e3689ad8>

> I _think_ none. I think this is fine.

Agree.

**Comment by hallyn** [comment]
I'm worried about that second one (https://github.com/shadow-maint/shadow/commit/928172c43495c08b8eb7502cd7890f78e3689ad8).

**Comment by alejandro-colomar** [comment]
> I'm worried about that second one ([928172c](https://github.com/shadow-maint/shadow/commit/928172c43495c08b8eb7502cd7890f78e3689ad8)).

That's the one I'm less worried about.  If they wanted to parse the output, they had the option to pass `-i` for a standard format.

I was more worried about the first one, since it was unconditionally non-standard.

(Or you might be worried about a script that would be written before chage(1) ever got a `-i` flag, which _might_ be a possibility.  Let's hope such a script doesn't exist.)

**Comment by hallyn** [comment]
@alejandro-colomar Ok, for all of the commits (including the already recently merged ones), let's please add an /etc/login.defs switch to dictate whether to use the classic format, or the standards-compliant formats.

I'm 100% convinced there's some site out there parsing the date on '-' (%F) or on the %b %d %Y, and https://github.com/shadow-maint/shadow/commit/928172c43495c08b8eb7502cd7890f78e3689ad8 will mess them up.

Generally I'd prefer having fewer ways of dictating the format as it's going to make it harder to spot an accidental format printing bug, but I think it's worth it here.

**Comment by alejandro-colomar** [comment]
> @alejandro-colomar Ok, for all of the commits (including the already recently merged ones), let's please add an /etc/login.defs switch to dictate whether to use the classic format, or the standards-compliant formats.
> 
> I'm 100% convinced there's some site out there parsing the date on '-' (%F) or on the %b %d %Y, and [928172c](https://github.com/shadow-maint/shadow/commit/928172c43495c08b8eb7502cd7890f78e3689ad8) will mess them up.
> 
> Generally I'd prefer having fewer ways of dictating the format as it's going to make it harder to spot an accidental format printing bug, but I think it's worth it here.

Hmm, I don't like adding configurability either.  How about the following instead?

Deprecate running passwd without -i.  Print a deprecation warning to stderr when a date is printed in the non-standard format.  Keep the nonstandard format untouched, for compatibility reasons. 

Then we can come back in 4 or 5 years and unconditionally enable %F.

**Comment by kenion** [comment]
Just for clarification, has a consensus been reached on how to move forward?  Since change #1058 was reverted, issue #1057 is still present.  For now, any user who updates to 4.14.6 from an earlier version will see the change in date input/output behavior described in that issue.

**Comment by alejandro-colomar** [comment]
Hi Gus,

On Mon, Aug 26, 2024 at 03:18:44AM GMT, Gus Kenion wrote:
> Just for clarification, has a consensus been reached on how to move
> forward?

Not yet.

> Since change #1058 was reverted,

No, it wasn't reverted.

> issue #1057 is still present.

And no, issue 1057 isn't present.

We have a different one now: we print a different date than the one that
is stored, since we print a local date, but store a UTC date.  This
might be problematic too.

The solution would be to append a Z to the ????-??-?? date and revert
#1058 to print a UTC date again.  However, that might break scripts, so
I would just leave it broken.  Anyone interested in the actual date,
please parse /etc/shadow themselves; that's certainly more precise.

Have a lovely day!
Alex

>  For now, any user who updates to 4.14.6 from an earlier version will
> see the change in date input/output behavior described in that issue.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/shadow-maint/shadow/pull/1059#issuecomment-2309862694
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

-- 
<https://www.alejandro-colomar.es/>

**Comment by kenion** [comment]
@alejandro-colomar sorry for my misunderstanding.  I've gone over the changes and comments again and I see the root cause of #1057 is addressed - thanks again to you and @hallyn.
